pkg/network/driver_interface.go

```diff
@@ -279,6 +351,45 @@ func resizeDriverBuffer(compareSize int, buffer []uint8) []uint8 {
 	return buffer
 }
 
+// ReadDNSPacket visits a raw DNS packet if one is available.
+func (di *DriverInterface) ReadDNSPacket(visit func([]byte, time.Time) error) (didRead bool, err error) {
+	var bytesRead uint32
+	var key uint32 // returned by GetQueuedCompletionStatus, then ignored
+	var ol *windows.Overlapped
+
+	// NOTE: ideally we would pass a timeout of INFINITY to the GetQueuedCompletionStatus, but are using a
+	// timeout of 0 and letting userspace do a busy loop to align better with the Linux code.
+	err = windows.GetQueuedCompletionStatus(di.dnsIOCP, &bytesRead, &key, &ol, 0)
+	if err != nil {
+		if err == syscall.Errno(syscall.WAIT_TIMEOUT) {
+			// this indicates that there was no queued completion status, this is fine
+			return false, nil
+		}
+
+		return false, errors.Wrap(err, "could not get queued completion status")
+	}
+
+	var buf *_readbuffer
+	buf = (*_readbuffer)(unsafe.Pointer(ol))
+
+	fph := (*C.struct_filterPacketHeader)(unsafe.Pointer(&buf.data[0]))
+	captureTime := time.Unix(0, int64(fph.timestamp))
+
+	start := C.sizeof_struct_filterPacketHeader
+
+	if err := visit(buf.data[start:], captureTime); err != nil {
+		return false, err
+	}
+
+	// kick off another read
+	if err := windows.ReadFile(di.driverDNSHandle.handle, buf.data[:], nil, &(buf.ol)); err != nil && err != windows.ERROR_IO_PENDING {
+		log.Errorf("%s", errors.Wrap(err, "cannot restart async"))
+		return false, err
```

Comment:
In general, you log OR return an error. Here we are doing both. Is there a reason?

Reply:
This was done to be consistent with the rest of the `driver_interface.go` file. We can clean up the logging in this file, but I would rather do that in a follow up PR.

Reply:
My concern is that we will get two log lines from the same error. This `ERROR` level one here, and then a `WARN` from https://github.com/DataDog/datadog-agent/blob/2417f917767e57520c02d79366ae8855bc22c1e3/pkg/network/dns_snooper.go#L184